File: Codigo/Nivel 1/Nivel 2/Nivel 3/Nivel 4/MTIS-Practica2-Proyecto/oficina1_java/src/mtis/Oficina1.java
package mtis;

import java.util.concurrent.CountDownLatch;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;
import javax.jms.Connection;
import javax.jms.ConnectionFactory;
import javax.jms.Destination;
import javax.jms.JMSException;
import javax.jms.MessageConsumer;
import javax.jms.MessageProducer;
import javax.jms.Session;
import javax.jms.TextMessage;
import javax.jms.Message;
import javax.jms.MessageListener;
import org.apache.activemq.ActiveMQConnectionFactory;
import org.json.JSONObject;

public class Oficina1 implements MessageListener {

    // Temperature variables 
    private int temperature;
    private boolean heatSystemActivated = false;
    private boolean coldSystemActivated = false;
    
    // Illumination variables
    private int illuminationIntensity;
    private boolean increaseIntensityIlluminationRegulatorActivated = false;
    private boolean decreaseIntensityIlluminationRegulatorActivated = false;

    public static final int TEMPERATURE_DIFFERENCE = 2;
    public static final int HEAT_SYSTEM_STOP_TEMPERATURE = 23;
    public static final int COLD_SYSTEM_STOP_TEMPERATURE = 23;
    
    public static final int ILLUMINATION_INTENSITY_DIFFERENCE = 200;
    public static final int INCREASE_INTENSITY_ILLUMINATION_REGULATOR_STOP_INTENSITY = 2300;
    public static final int DECREASE_INTENSITY_ILLUMINATION_REGULATOR_STOP_INTENSITY = 2300;

    // Constructor: initialize temperature and illumination intensity
    public Oficina1() {
        this.temperature = 0; // initial temperature value
        this.illuminationIntensity = 2300; // initial illumination intensity (equilibrium)
    }

    public int getTemperature() {
        return temperature;
    }

    public void setTemperature(int temperature) {
        this.temperature = temperature;
    }
    
    public boolean isColdSystemActivated() {
        return coldSystemActivated;
    }

    public void setColdSystemActivated(boolean coldSystemActivated) {
        this.coldSystemActivated = coldSystemActivated;
    }

    public boolean isHeatSystemActivated() {
        return heatSystemActivated;
    }

    public void setHeatSystemActivated(boolean heatSystemActivated) {
        this.heatSystemActivated = heatSystemActivated;
    }
    
    public int getIlluminationIntensity() {
        return illuminationIntensity;
    }
    
    public void setIlluminationIntensity(int illuminationIntensity) {
        this.illuminationIntensity = illuminationIntensity;
    }

    public void incrementTemperature() {
        this.temperature += TEMPERATURE_DIFFERENCE;
    }

    public void decrementTemperature() {
        this.temperature -= TEMPERATURE_DIFFERENCE;
    }

    /**
     * Internal class to hold JMS components for both temperature and illumination.
     */
    private static class JMSComponents {
        public Connection connection;
        public Session session;
        public MessageProducer temperatureProducer;
        public MessageConsumer temperatureConsumer;
        public MessageProducer illuminationProducer;
        public MessageConsumer illuminationConsumer;
        
        public JMSComponents(Connection connection, Session session,
                             MessageProducer temperatureProducer, MessageConsumer temperatureConsumer,
                             MessageProducer illuminationProducer, MessageConsumer illuminationConsumer) {
            this.connection = connection;
            this.session = session;
            this.temperatureProducer = temperatureProducer;
            this.temperatureConsumer = temperatureConsumer;
            this.illuminationProducer = illuminationProducer;
            this.illuminationConsumer = illuminationConsumer;
        }
    }
    
    /**
     * Setup JMS connection, session, and destinations for both temperature and illumination.
     */
    private JMSComponents setupJMS() throws JMSException {
        System.out.println("Office1 Starting... ");
        String url = "tcp://localhost:61616";
        
        CountDownLatch latch = new CountDownLatch(1);
        
        // Create connection and session  
        ConnectionFactory connectionFactory = new ActiveMQConnectionFactory("mtis", "mtis", url);
        Connection connection = connectionFactory.createConnection();
        connection.start();
        Session session = connection.createSession(false, Session.AUTO_ACKNOWLEDGE);
        
        // Create destinations for temperature
        Destination lecturasTempDest = session.createTopic("lecturas_temperaturas_oficina1");
        Destination actuadorTempDest = session.createTopic("actuador_temperatura_oficina1");
        
        // Create destinations for illumination
        Destination lecturasIllumDest = session.createTopic("lecturas_iluminacion_oficina1");
        Destination actuadorIllumDest = session.createTopic("actuador_iluminacion_oficina1");
        
        // Create producer and consumer for temperature
        MessageProducer temperatureProducer = session.createProducer(lecturasTempDest);
        MessageConsumer temperatureConsumer = session.createConsumer(actuadorTempDest);
        temperatureConsumer.setMessageListener(this);
        
        // Create producer and consumer for illumination
        MessageProducer illuminationProducer = session.createProducer(lecturasIllumDest);
        MessageConsumer illuminationConsumer = session.createConsumer(actuadorIllumDest);
        illuminationConsumer.setMessageListener(this);
        
        return new JMSComponents(connection, session, temperatureProducer, temperatureConsumer, illuminationProducer, illuminationConsumer);
    }
    
    /**
     * Sends a temperature reading message.
     */
    private void sendTemperatureMessage(Session session, MessageProducer producer) throws JMSException {
        String jsonPayload = "{"
                + "\"office\": \"office1\","
                + "\"temperature\": " + this.temperature + ","
                + "\"cold_system_activated\": " + this.coldSystemActivated + ","
                + "\"heat_system_activated\": " + this.heatSystemActivated
                + "}";
        TextMessage message = session.createTextMessage(jsonPayload);
        producer.send(message);
    }

    /**
     * Manages the cold system: decrements temperature, sends message, and deactivates if condition met.
     */
    public void manageColdSystem(Session session, MessageProducer producer) throws JMSException {
        if (!this.coldSystemActivated) {
            this.setColdSystemActivated(true);
        }
        this.decrementTemperature();
        
        String jsonPayload = "{"
                + "\"office\": \"office1\","
                + "\"temperature\": " + this.getTemperature() + ","
                + "\"cold_system_activated\": " + this.coldSystemActivated + ","
                + "\"heat_system_activated\": " + this.heatSystemActivated
                + "}";
        
        TextMessage message = session.createTextMessage(jsonPayload);
        producer.send(message);
        
        if (this.getTemperature() <= COLD_SYSTEM_STOP_TEMPERATURE) {
            this.setColdSystemActivated(false);
        }
    }

    /**
     * Manages the heat system: increments temperature, sends message, and deactivates if condition met.
     */
    public void manageHeatSystem(Session session, MessageProducer producer) throws JMSException {
        if (!this.heatSystemActivated) {
            this.setHeatSystemActivated(true);
        }
        this.incrementTemperature();
        
        String jsonPayload = "{"
                + "\"office\": \"office1\","
                + "\"temperature\": " + this.getTemperature() + ","
                + "\"cold_system_activated\": " + this.coldSystemActivated + ","
                + "\"heat_system_activated\": " + this.heatSystemActivated
                + "}";
        
        TextMessage message = session.createTextMessage(jsonPayload);
        producer.send(message);
        
        if (this.getTemperature() >= HEAT_SYSTEM_STOP_TEMPERATURE) {
            this.setHeatSystemActivated(false);
        }
    }

    // Temperature flag management methods (unchanged)
    public Boolean getActivateColdSystemFlag(String text) {
        JSONObject json = new JSONObject(text);
        if (json.isNull("activate_cold_system")) {
            return null;
        }
        else {
            return json.getBoolean("activate_cold_system");
        }
    }

    public Boolean getStopColdSystemFlag(String text) {
        JSONObject json = new JSONObject(text);
        if (json.isNull("stop_cold_system")) {
            return null;
        } else {
            return json.getBoolean("stop_cold_system");
        }
    }

    public Boolean getActivateHeatSystemFlag(String text) {
        JSONObject json = new JSONObject(text);
        if (json.isNull("activate_heat_system")) {
            return null;
        } else {
            return json.getBoolean("activate_heat_system");
        }
    }

    public Boolean getStopHeatSystemFlag(String text) {
        JSONObject json = new JSONObject(text);
        if (json.isNull("stop_heat_system")) {
            return null;
        } else {
            return json.getBoolean("stop_heat_system");
        }
    }

    public void manageTemperatureFlags(String text) {
        Boolean activateColdSystem = getActivateColdSystemFlag(text);
        Boolean stopColdSystem = getStopColdSystemFlag(text);
        Boolean activateHeatSystem = getActivateHeatSystemFlag(text);
        Boolean stopHeatSystem = getStopHeatSystemFlag(text);

        if (activateColdSystem != null && activateColdSystem) {
            this.setColdSystemActivated(true);
        }
        if (stopColdSystem != null && stopColdSystem) {
            this.setColdSystemActivated(false);
        }
        if (activateHeatSystem != null && activateHeatSystem) {
            this.setHeatSystemActivated(true);
        }
        if (stopHeatSystem != null && stopHeatSystem) {
            this.setHeatSystemActivated(false);
        }
    }

    public void printOwnTemperatureInformation() {
        System.out.println("Temperature Sensor: " + this.getTemperature() + "C");
        if (this.isColdSystemActivated()) {
            System.out.println("Temperature Activator: Cold System activated.");
        }
        if (this.isHeatSystemActivated()) {
            System.out.println("Temperature Activator: Heat System activated.");
        }
    }

    // New methods for illumination

    private void sendIlluminationMessage(Session session, MessageProducer producer) throws JMSException {
        String jsonPayload = "{"
                + "\"office\": \"office1\","
                + "\"illumination_intensity\": " + this.illuminationIntensity + ","
                + "\"increase_intensity_regulator_activated\": " + this.increaseIntensityIlluminationRegulatorActivated + ","
                + "\"decrease_intensity_regulator_activated\": " + this.decreaseIntensityIlluminationRegulatorActivated
                + "}";
        TextMessage message = session.createTextMessage(jsonPayload);
        producer.send(message);
    }

    public void manageIncreaseIlluminationRegulator(Session session, MessageProducer producer) throws JMSException {
        if (!this.increaseIntensityIlluminationRegulatorActivated) {
            this.increaseIntensityIlluminationRegulatorActivated = true;
        }
        this.illuminationIntensity += ILLUMINATION_INTENSITY_DIFFERENCE;
        
        String jsonPayload = "{"
                + "\"office\": \"office1\","
                + "\"illumination_intensity\": " + this.illuminationIntensity + ","
                + "\"increase_intensity_regulator_activated\": " + this.increaseIntensityIlluminationRegulatorActivated + ","
                + "\"decrease_intensity_regulator_activated\": " + this.decreaseIntensityIlluminationRegulatorActivated
                + "}";
        TextMessage message = session.createTextMessage(jsonPayload);
        producer.send(message);
        
        if (this.illuminationIntensity >= INCREASE_INTENSITY_ILLUMINATION_REGULATOR_STOP_INTENSITY) {
            this.increaseIntensityIlluminationRegulatorActivated = false;
        }
    }

    public void manageDecreaseIlluminationRegulator(Session session, MessageProducer producer) throws JMSException {
        if (!this.decreaseIntensityIlluminationRegulatorActivated) {
            this.decreaseIntensityIlluminationRegulatorActivated = true;
        }
        this.illuminationIntensity -= ILLUMINATION_INTENSITY_DIFFERENCE;
        
        String jsonPayload = "{"
                + "\"office\": \"office1\","
                + "\"illumination_intensity\": " + this.illuminationIntensity + ","
                + "\"increase_intensity_regulator_activated\": " + this.increaseIntensityIlluminationRegulatorActivated + ","
                + "\"decrease_intensity_regulator_activated\": " + this.decreaseIntensityIlluminationRegulatorActivated
                + "}";
        TextMessage message = session.createTextMessage(jsonPayload);
        producer.send(message);
        
        if (this.illuminationIntensity <= DECREASE_INTENSITY_ILLUMINATION_REGULATOR_STOP_INTENSITY) {
            this.decreaseIntensityIlluminationRegulatorActivated = false;
        }
    }

    public Boolean getActivateIncreaseIntensityFlag(String text) {
        JSONObject json = new JSONObject(text);
        if (json.isNull("activate_increase_intensity_illumination_regulator")) {
            return null;
        } else {
            return json.getBoolean("activate_increase_intensity_illumination_regulator");
        }
    }

    public Boolean getStopIncreaseIntensityFlag(String text) {
        JSONObject json = new JSONObject(text);
        if (json.isNull("stop_increase_intensity_illumination_regulator")) {
            return null;
        } else {
            return json.getBoolean("stop_increase_intensity_illumination_regulator");
        }
    }

    public Boolean getActivateDecreaseIntensityFlag(String text) {
        JSONObject json = new JSONObject(text);
        if (json.isNull("activate_decrease_intensity_illumination_regulator")) {
            return null;
        } else {
            return json.getBoolean("activate_decrease_intensity_illumination_regulator");
        }
    }

    public Boolean getStopDecreaseIntensityFlag(String text) {
        JSONObject json = new JSONObject(text);
        if (json.isNull("stop_decrease_intensity_illumination_regulator")) {
            return null;
        } else {
            return json.getBoolean("stop_decrease_intensity_illumination_regulator");
        }
    }

    public void manageIlluminationFlags(String text) {
        Boolean activateIncrease = getActivateIncreaseIntensityFlag(text);
        Boolean stopIncrease = getStopIncreaseIntensityFlag(text);
        Boolean activateDecrease = getActivateDecreaseIntensityFlag(text);
        Boolean stopDecrease = getStopDecreaseIntensityFlag(text);

        if (activateIncrease != null && activateIncrease) {
            this.increaseIntensityIlluminationRegulatorActivated = true;
        }
        if (stopIncrease != null && stopIncrease) {
            this.increaseIntensityIlluminationRegulatorActivated = false;
        }
        if (activateDecrease != null && activateDecrease) {
            this.decreaseIntensityIlluminationRegulatorActivated = true;
        }
        if (stopDecrease != null && stopDecrease) {
            this.decreaseIntensityIlluminationRegulatorActivated = false;
        }
    }

    public void printOwnIlluminationInformation() {
        System.out.println("Illumination Sensor: " + this.getIlluminationIntensity() + " lumens");
        if (this.increaseIntensityIlluminationRegulatorActivated) {
            System.out.println("Illumination Activator: Increase Intensity Regulator activated.");
        }
        if (this.decreaseIntensityIlluminationRegulatorActivated) {
            System.out.println("Illumination Activator: Decrease Intensity Regulator activated.");
        }
    }

    public void printReceivedIlluminationInformation(String text) {
        if (getActivateIncreaseIntensityFlag(text) != null && getActivateIncreaseIntensityFlag(text)) {
            System.out.println("Illumination Activator: Illumination intensity is below threshold (below 1500 lumens). Activating Increase Intensity Regulator...");
        }
        if (getStopIncreaseIntensityFlag(text) != null && getStopIncreaseIntensityFlag(text)) {
            System.out.println("Illumination Activator: Illumination intensity has reached desired level (around 2300 lumens). Stopping Increase Intensity Regulator...");
        }
        if (getActivateDecreaseIntensityFlag(text) != null && getActivateDecreaseIntensityFlag(text)) {
            System.out.println("Illumination Activator: Illumination intensity is above threshold (above 3000 lumens). Activating Decrease Intensity Regulator...");
        }
        if (getStopDecreaseIntensityFlag(text) != null && getStopDecreaseIntensityFlag(text)) {
            System.out.println("Illumination Activator: Illumination intensity has reached desired level (around 2300 lumens). Stopping Decrease Intensity Regulator...");
        }
    }
    
    /**
     * Modified onMessage to process both temperature and illumination messages.
     */
    @Override
    public void onMessage(Message message) {
        try {
            if (message instanceof TextMessage) {
                TextMessage textMessage = (TextMessage) message;
                String text = textMessage.getText();
                JSONObject json = new JSONObject(text);
                if (json.has("temperature")) {
                    manageTemperatureFlags(text);
                    printReceivedTemperatureInformation(text);
                }
                if (json.has("illumination_intensity")) {
                    manageIlluminationFlags(text);
                    printReceivedIlluminationInformation(text);
                }
            } else {
                System.out.println("Received message of type: " + message.getClass().getName());
            }
        } catch (JMSException e) {
            System.out.println("Got a JMS Exception!");
        }
    }
    
    /**
     * The start method now schedules two tasks: one for temperature and one for illumination.
     */
    public void start() throws JMSException {
        final JMSComponents jmsComponents = setupJMS();
        
        final ScheduledExecutorService executor = Executors.newScheduledThreadPool(4);
        
        // Temperature task
        executor.scheduleAtFixedRate(new Runnable() {
            @Override
            public void run() {
                try {
                    if (!Oficina1.this.isColdSystemActivated() && !Oficina1.this.isHeatSystemActivated()) {
                        Oficina1.this.setTemperature(Utils.manejarTemperaturaRandomIndicator());
                    }
                    if (Oficina1.this.isHeatSystemActivated()) {
                        Oficina1.this.manageHeatSystem(jmsComponents.session, jmsComponents.temperatureProducer);
                    } else if (Oficina1.this.isColdSystemActivated()) {
                        Oficina1.this.manageColdSystem(jmsComponents.session, jmsComponents.temperatureProducer);
                    } else {
                        Oficina1.this.sendTemperatureMessage(jmsComponents.session, jmsComponents.temperatureProducer);
                    }
                    Oficina1.this.printOwnTemperatureInformation();
                } catch(Exception e) {
                    e.printStackTrace();
                }
            }
        }, 0, 2, TimeUnit.SECONDS);
        
        // Illumination task
        executor.scheduleAtFixedRate(new Runnable() {
            @Override
            public void run() {
                try {
                    if (!Oficina1.this.increaseIntensityIlluminationRegulatorActivated && !Oficina1.this.decreaseIntensityIlluminationRegulatorActivated) {
                        Oficina1.this.setIlluminationIntensity(Utils.manejarIlluminationIntensityRandomIndicator());
                    }
                    if (Oficina1.this.increaseIntensityIlluminationRegulatorActivated) {
                        Oficina1.this.manageIncreaseIlluminationRegulator(jmsComponents.session, jmsComponents.illuminationProducer);
                    } else if (Oficina1.this.decreaseIntensityIlluminationRegulatorActivated) {
                        Oficina1.this.manageDecreaseIlluminationRegulator(jmsComponents.session, jmsComponents.illuminationProducer);
                    } else {
                        Oficina1.this.sendIlluminationMessage(jmsComponents.session, jmsComponents.illuminationProducer);
                    }
                    Oficina1.this.printOwnIlluminationInformation();
                } catch(Exception e) {
                    e.printStackTrace();
                }
            }
        }, 0, 2, TimeUnit.SECONDS);
    }
    
    public void printReceivedTemperatureInformation(String text) {
        if (this.getActivateColdSystemFlag(text) != null && this.getActivateColdSystemFlag(text)) {
            System.out.println("Temperature Activator: Temperature has exceeded 30C. Activating Cold System...");
        }
        if (this.getStopColdSystemFlag(text) != null && this.getStopColdSystemFlag(text)) {
            System.out.println("Temperature Activator: Temperature has reached 23C or less. Stopping Cold System...");
        }
        if (this.getActivateHeatSystemFlag(text) != null && this.getActivateHeatSystemFlag(text)) {
            System.out.println("Temperature Activator: Temperature is below 15C. Activating Heat System...");
        }
        if (this.getStopHeatSystemFlag(text) != null && this.getStopHeatSystemFlag(text)) {
            System.out.println("Temperature Activator: Temperature has reached 23C or more. Stopping Heat System...");
        }
    }
    
    public static void main(String[] args) throws JMSException {
        Oficina1 oficina = new Oficina1();
        oficina.start();
    }
}

File: Codigo/Nivel 1/Nivel 2/Nivel 3/Nivel 4/MTIS-Practica2-Proyecto/oficina1_php/Oficina1.php
<?php
// Oficina1.php
// PHP-based replacement for the Java "Oficina1" using stomp-php 5.1.2 and react/event-loop
//
// Make sure to run "composer install" so that the following classes are available:
//   - Stomp\Client
//   - Stomp\StatefulStomp
//   - React\EventLoop\Factory
// Then run this script via: php Oficina1.php

require_once __DIR__ . '/vendor/autoload.php';
require_once __DIR__ . '/Utils.php';

use Stomp\Client;
use Stomp\StatefulStomp;
use Stomp\Transport\Message;
use React\EventLoop\Factory;
use Oficina1Utils\Utils;

class Oficina1
{
    // Temperature variables
    private $temperature;
    private $heatSystemActivated = false;
    private $coldSystemActivated = false;

    // Illumination variables
    private $illuminationIntensity;
    private $increaseIntensityIlluminationRegulatorActivated = false;
    private $decreaseIntensityIlluminationRegulatorActivated = false;

    // Constants analogous to Java
    const TEMPERATURE_DIFFERENCE = 2;
    const HEAT_SYSTEM_STOP_TEMPERATURE = 23;
    const COLD_SYSTEM_STOP_TEMPERATURE = 23;

    const ILLUMINATION_INTENSITY_DIFFERENCE = 200;
    const INCREASE_INTENSITY_ILLUMINATION_REGULATOR_STOP_INTENSITY = 2300;
    const DECREASE_INTENSITY_ILLUMINATION_REGULATOR_STOP_INTENSITY = 2300;

    // Topics
    private $topicLecturasTemperature = '/topic/lecturas_temperaturas_oficina1';
    private $topicActuadorTemperature = '/topic/actuador_temperatura_oficina1';
    private $topicLecturasIllumination = '/topic/lecturas_iluminacion_oficina1';
    private $topicActuadorIllumination = '/topic/actuador_iluminacion_oficina1';

    /** @var StatefulStomp */
    private $stomp;
    /** @var \React\EventLoop\LoopInterface */
    private $loop;

    public function __construct()
    {
        // same initial values as in Oficina1.java
        $this->temperature = 0;
        $this->illuminationIntensity = 2300;
    }

    public function start()
    {
        echo "Office1 Starting... \n";

        $this->loop = Factory::create();

        // 1) Create the stomp client
        $client = new Client('tcp://localhost:61613');
        $client->setLogin('mtis', 'mtis'); // user + pass
        $client->connect();
        // 2) Wrap it in a StatefulStomp
        $this->stomp = new StatefulStomp($client);

        // 3) Subscribe to the actuator topics
        //    We'll poll them in the event loop to keep it asynchronous
        $this->stomp->subscribe($this->topicActuadorTemperature, null, 'auto');
        $this->stomp->subscribe($this->topicActuadorIllumination, null, 'auto');

        // Continuously read frames in a small periodic timer
        $this->loop->addPeriodicTimer(0.1, function () {
            if ($frame = $this->stomp->read()) {
                // We have a message
                $this->onMessage($frame);
            }
        });

        // Define temperature task
        $temperatureTask = function () {
            // If no system is active, randomize temperature
            if (!$this->coldSystemActivated && !$this->heatSystemActivated) {
                $this->temperature = Utils::manejarTemperaturaRandomIndicator();
            }
            if ($this->heatSystemActivated) {
                $this->manageHeatSystem();
            } elseif ($this->coldSystemActivated) {
                $this->manageColdSystem();
            } else {
                // Just send the current temperature
                $this->sendTemperatureMessage();
            }
            $this->printOwnTemperatureInformation();
        };

        // Define illumination task
        $illuminationTask = function () {
            // If no regulator is active, randomize illumination
            if (!$this->increaseIntensityIlluminationRegulatorActivated && !$this->decreaseIntensityIlluminationRegulatorActivated) {
                $this->illuminationIntensity = Utils::manejarIlluminationIntensityRandomIndicator();
            }
            if ($this->increaseIntensityIlluminationRegulatorActivated) {
                $this->manageIncreaseIlluminationRegulator();
            } elseif ($this->decreaseIntensityIlluminationRegulatorActivated) {
                $this->manageDecreaseIlluminationRegulator();
            } else {
                // Just send the current illumination
                $this->sendIlluminationMessage();
            }
            $this->printOwnIlluminationInformation();
        };

        // Execute tasks immediately so messages appear without delay
        $temperatureTask();
        $illuminationTask();

        // Schedule temperature task every 2 seconds
        $this->loop->addPeriodicTimer(2.0, $temperatureTask);

        // Schedule illumination task every 2 seconds
        $this->loop->addPeriodicTimer(2.0, $illuminationTask);

        // Run the React event loop
        $this->loop->run();
    }

    // ---- JMS-like onMessage (actually a frame read) ----
    private function onMessage($frame)
    {
        $body = $frame->getBody();
        if (empty($body)) {
            echo "Received empty message.\n";
            return;
        }
        // parse as JSON
        $data = json_decode($body, true);
        if ($data === null) {
            echo "Got invalid JSON.\n";
            return;
        }

        // If there's a "temperature" field, handle temperature flags
        if (array_key_exists('temperature', $data)) {
            $this->manageTemperatureFlags($data);
            $this->printReceivedTemperatureInformation($data);
        }
        // If there's an "illumination_intensity" field, handle illumination
        if (array_key_exists('illumination_intensity', $data)) {
            $this->manageIlluminationFlags($data);
            $this->printReceivedIlluminationInformation($data);
        }
    }

    // ---- Manage temperature flags ----
    private function manageTemperatureFlags($json)
    {
        if (isset($json['activate_cold_system']) && $json['activate_cold_system'] === true) {
            $this->coldSystemActivated = true;
        }
        if (isset($json['stop_cold_system']) && $json['stop_cold_system'] === true) {
            $this->coldSystemActivated = false;
        }
        if (isset($json['activate_heat_system']) && $json['activate_heat_system'] === true) {
            $this->heatSystemActivated = true;
        }
        if (isset($json['stop_heat_system']) && $json['stop_heat_system'] === true) {
            $this->heatSystemActivated = false;
        }
    }

    // ---- Manage illumination flags ----
    private function manageIlluminationFlags($json)
    {
        if (isset($json['activate_increase_intensity_illumination_regulator']) 
            && $json['activate_increase_intensity_illumination_regulator'] === true) {
            $this->increaseIntensityIlluminationRegulatorActivated = true;
        }
        if (isset($json['stop_increase_intensity_illumination_regulator']) 
            && $json['stop_increase_intensity_illumination_regulator'] === true) {
            $this->increaseIntensityIlluminationRegulatorActivated = false;
        }
        if (isset($json['activate_decrease_intensity_illumination_regulator']) 
            && $json['activate_decrease_intensity_illumination_regulator'] === true) {
            $this->decreaseIntensityIlluminationRegulatorActivated = true;
        }
        if (isset($json['stop_decrease_intensity_illumination_regulator']) 
            && $json['stop_decrease_intensity_illumination_regulator'] === true) {
            $this->decreaseIntensityIlluminationRegulatorActivated = false;
        }
    }

    // ---- Methods for incrementing/decrementing temperature ----
    private function incrementTemperature()
    {
        $this->temperature += self::TEMPERATURE_DIFFERENCE;
    }

    private function decrementTemperature()
    {
        $this->temperature -= self::TEMPERATURE_DIFFERENCE;
    }

    // ---- Send or manage temperature messages ----
    private function sendTemperatureMessage()
    {
        $payloadArray = [
            "office"                => "office1",
            "temperature"           => $this->temperature,
            "cold_system_activated" => $this->coldSystemActivated,
            "heat_system_activated" => $this->heatSystemActivated
        ];
        $payloadJson = json_encode($payloadArray);
        $msg = new Message($payloadJson, ["content-type" => "application/json"]);
        $this->stomp->send($this->topicLecturasTemperature, $msg);
    }

    public function manageColdSystem()
    {
        if (!$this->coldSystemActivated) {
            $this->coldSystemActivated = true;
        }
        $this->decrementTemperature();

        $payloadArray = [
            "office"                => "office1",
            "temperature"           => $this->temperature,
            "cold_system_activated" => $this->coldSystemActivated,
            "heat_system_activated" => $this->heatSystemActivated
        ];
        $payloadJson = json_encode($payloadArray);
        $msg = new Message($payloadJson, ["content-type" => "application/json"]);
        $this->stomp->send($this->topicLecturasTemperature, $msg);

        if ($this->temperature <= self::COLD_SYSTEM_STOP_TEMPERATURE) {
            $this->coldSystemActivated = false;
        }
    }

    public function manageHeatSystem()
    {
        if (!$this->heatSystemActivated) {
            $this->heatSystemActivated = true;
        }
        $this->incrementTemperature();

        $payloadArray = [
            "office"                => "office1",
            "temperature"           => $this->temperature,
            "cold_system_activated" => $this->coldSystemActivated,
            "heat_system_activated" => $this->heatSystemActivated
        ];
        $payloadJson = json_encode($payloadArray);
        $msg = new Message($payloadJson, ["content-type" => "application/json"]);
        $this->stomp->send($this->topicLecturasTemperature, $msg);

        if ($this->temperature >= self::HEAT_SYSTEM_STOP_TEMPERATURE) {
            $this->heatSystemActivated = false;
        }
    }

    public function printOwnTemperatureInformation()
    {
        echo "Temperature Sensor: " . $this->temperature . "C\n";
        if ($this->coldSystemActivated) {
            echo "Temperature Activator: Cold System activated.\n";
        }
        if ($this->heatSystemActivated) {
            echo "Temperature Activator: Heat System activated.\n";
        }
    }

    public function printReceivedTemperatureInformation($json)
    {
        if (isset($json['activate_cold_system']) && $json['activate_cold_system'] === true) {
            echo "Temperature Activator: Temperature has exceeded 30C. Activating Cold System...\n";
        }
        if (isset($json['stop_cold_system']) && $json['stop_cold_system'] === true) {
            echo "Temperature Activator: Temperature has reached 23C or less. Stopping Cold System...\n";
        }
        if (isset($json['activate_heat_system']) && $json['activate_heat_system'] === true) {
            echo "Temperature Activator: Temperature is below 15C. Activating Heat System...\n";
        }
        if (isset($json['stop_heat_system']) && $json['stop_heat_system'] === true) {
            echo "Temperature Activator: Temperature has reached 23C or more. Stopping Heat System...\n";
        }
    }

    // ---- Methods for illumination ----
    private function sendIlluminationMessage()
    {
        $payloadArray = [
            "office"                                 => "office1",
            "illumination_intensity"                 => $this->illuminationIntensity,
            "increase_intensity_regulator_activated" => $this->increaseIntensityIlluminationRegulatorActivated,
            "decrease_intensity_regulator_activated" => $this->decreaseIntensityIlluminationRegulatorActivated
        ];
        $payloadJson = json_encode($payloadArray);
        $msg = new Message($payloadJson, ["content-type" => "application/json"]);
        $this->stomp->send($this->topicLecturasIllumination, $msg);
    }

    public function manageIncreaseIlluminationRegulator()
    {
        if (!$this->increaseIntensityIlluminationRegulatorActivated) {
            $this->increaseIntensityIlluminationRegulatorActivated = true;
        }
        $this->illuminationIntensity += self::ILLUMINATION_INTENSITY_DIFFERENCE;

        $payloadArray = [
            "office"                                 => "office1",
            "illumination_intensity"                 => $this->illuminationIntensity,
            "increase_intensity_regulator_activated" => $this->increaseIntensityIlluminationRegulatorActivated,
            "decrease_intensity_regulator_activated" => $this->decreaseIntensityIlluminationRegulatorActivated
        ];
        $payloadJson = json_encode($payloadArray);
        $msg = new Message($payloadJson, ["content-type" => "application/json"]);
        $this->stomp->send($this->topicLecturasIllumination, $msg);

        if ($this->illuminationIntensity >= self::INCREASE_INTENSITY_ILLUMINATION_REGULATOR_STOP_INTENSITY) {
            $this->increaseIntensityIlluminationRegulatorActivated = false;
        }
    }

    public function manageDecreaseIlluminationRegulator()
    {
        if (!$this->decreaseIntensityIlluminationRegulatorActivated) {
            $this->decreaseIntensityIlluminationRegulatorActivated = true;
        }
        $this->illuminationIntensity -= self::ILLUMINATION_INTENSITY_DIFFERENCE;

        $payloadArray = [
            "office"                                 => "office1",
            "illumination_intensity"                 => $this->illuminationIntensity,
            "increase_intensity_regulator_activated" => $this->increaseIntensityIlluminationRegulatorActivated,
            "decrease_intensity_regulator_activated" => $this->decreaseIntensityIlluminationRegulatorActivated
        ];
        $payloadJson = json_encode($payloadArray);
        $msg = new Message($payloadJson, ["content-type" => "application/json"]);
        $this->stomp->send($this->topicLecturasIllumination, $msg);

        if ($this->illuminationIntensity <= self::DECREASE_INTENSITY_ILLUMINATION_REGULATOR_STOP_INTENSITY) {
            $this->decreaseIntensityIlluminationRegulatorActivated = false;
        }
    }

    public function printOwnIlluminationInformation()
    {
        echo "Illumination Sensor: " . $this->illuminationIntensity . " lumens\n";
        if ($this->increaseIntensityIlluminationRegulatorActivated) {
            echo "Illumination Activator: Increase Intensity Regulator activated.\n";
        }
        if ($this->decreaseIntensityIlluminationRegulatorActivated) {
            echo "Illumination Activator: Decrease Intensity Regulator activated.\n";
        }
    }

    public function printReceivedIlluminationInformation($json)
    {
        if (isset($json['activate_increase_intensity_illumination_regulator'])
            && $json['activate_increase_intensity_illumination_regulator'] === true) {
            echo "Illumination Activator: Illumination intensity is below threshold (below 1500 lumens). Activating Increase Intensity Regulator...\n";
        }
        if (isset($json['stop_increase_intensity_illumination_regulator'])
            && $json['stop_increase_intensity_illumination_regulator'] === true) {
            echo "Illumination Activator: Illumination intensity has reached desired level (around 2300 lumens). Stopping Increase Intensity Regulator...\n";
        }
        if (isset($json['activate_decrease_intensity_illumination_regulator'])
            && $json['activate_decrease_intensity_illumination_regulator'] === true) {
            echo "Illumination Activator: Illumination intensity is above threshold (above 3000 lumens). Activating Decrease Intensity Regulator...\n";
        }
        if (isset($json['stop_decrease_intensity_illumination_regulator'])
            && $json['stop_decrease_intensity_illumination_regulator'] === true) {
            echo "Illumination Activator: Illumination intensity has reached desired level (around 2300 lumens). Stopping Decrease Intensity Regulator...\n";
        }
    }
}

// ---- Entry point ---- 
$oficina = new Oficina1();
$oficina->start();
